crates/typst-cli/src/download.rs

```diff
@@ -0,0 +1,204 @@
+use std::collections::VecDeque;
+use std::io::{self, ErrorKind, Read, Stdout, Write};
+use std::time::{Duration, Instant};
+
+use ureq::Response;
+
+/// Keep track of this many past download amounts
+const DOWNLOAD_TRACK_COUNT: usize = 5;
+
+/// A wrapper around [`ureq::Response`] that reads the response body in chunks
+/// over a websocket and displays statistics about its progress.
+///
+/// Downloads will _never_ fail due to statistics failing to print, print errors
+/// are silently ignored.
+///
+/// [`ureq::Response`]: https://docs.rs/ureq/2.7.1/ureq/struct.Response.html
+pub struct RemoteReader {
+    reader: Box<dyn Read + Send + Sync + 'static>,
+    content_len: usize,
+    total_downloaded: usize,
+    downloaded_this_sec: usize,
+    downloaded_last_few_secs: VecDeque<usize>,
+    start_time: Instant,
+    last_print: Option<Instant>,
+    displayed_charcount: Option<usize>,
+    terminal: Stdout,
+}
+
+impl RemoteReader {
+    /// Wraps a [`ureq::Response`] and prepares it for downloading.
+    ///
+    /// Downloading data in chunks requires that the content length is known
+    /// before any reads take place. The response _must_ have a 'Content-Length'
+    /// header attached, otherwise downloading will fail.
```

Comment:
This does not seem ideal. Couldn't we just not print the total if the content length is unknown?

Reply:
I'll make it optional, so if the content length header isn't present it will just print the following:
```
4.6 MiB downloaded   1.1 MiB/s in  4s
```

Reply:
Now that I look into it again, we need to allocate a buffer large enough to hold data, we can't dynamically grow it:
`let mut data = vec![0; self.content_len];`, this was also my initial reasoning to let it fail if there was no content length header.

Reply:
Why can't you dynamically grow it? `self.reader.read(&mut data[offset..])` will read at most as many bytes as the slice can hold. If you don't know the content size, you can allocate a reasonable default size. When `read` fills the whole buffer, you increase the vector's size and continue with the loop to retrieve more bytes. 

You can see how `read_to_end` does this without knowing the size in advanced [here](https://doc.rust-lang.org/src/std/io/mod.rs.html#360). It uses a quite complex scheme for optimal efficiency, but we can do something simpler, e.g. just doubling the buffer if it's full. std uses a probing buffer to prevent doubling if the allocation was exactly correct, but you could also just overallocate a bit. Or stop reading if you reached the Content-Length. At the end of the code, you can call `data.truncate(offset)` to get rid of the excess without reallocating again.

From a quick look at the code, it also seems that your code panics if the final buffer size doesn't match the Content-Length. That's not ideal. Ideally, the Content-Length would be used as a hint to allocate enough space, but if it was incorrect, it doesn't matter.

Reply:
I think you were writing up a response while I already replied to myself, I thought of this:

```rust
let mut buffer = [0; 8192];
let mut data = match self.content_len {
    Some(content_len) => Vec::with_capacity(content_len),
    None => Vec::with_capacity(8192)
};

let read = match self.reader.read(&mut buffer) { .. }

data.extend(&buffer[..read]);
```

What do you think of that solution? `data` will eventually yield all bytes, it will reallocate a bunch though but I think it doesn't matter too much in our use-case right?

Reply:
Good catch, I forgot to add `vec!`. I'll push this solution shortly.